driver-sync/src/main/com/mongodb/client/internal/ClientEncryptionImpl.java

```diff
@@ -171,6 +183,65 @@ public RewrapManyDataKeyResult rewrapManyDataKey(final Bson filter, final Rewrap
+        MongoNamespace namespace = new MongoNamespace(database.getName(), collectionName);
+        Bson rawEncryptedFields = createCollectionOptions.getEncryptedFields();
+        if (rawEncryptedFields == null) {
+            throw new MongoConfigurationException(format("`encryptedFields` is not configured for the collection %s.", namespace));
+        }
+        CodecRegistry codecRegistry = options.getKeyVaultMongoClientSettings() == null
+                ? MongoClientSettings.getDefaultCodecRegistry()
+                : options.getKeyVaultMongoClientSettings().getCodecRegistry();
+        BsonDocument actualEncryptedFields = rawEncryptedFields.toBsonDocument(BsonDocument.class, codecRegistry).clone();
+        BsonValue fields = actualEncryptedFields.get("fields");
+        if (fields != null) {
+            if (!fields.isArray()) {
+                throw new MongoConfigurationException(format("`encryptedFields` is incorrectly configured for the collection %s."
+                        + " `encryptedFields.fields` must be an array, but is %s type.", namespace, fields.getBsonType()));
+            }
+            String kmsProvider = createEncryptedCollectionParams.getKmsProvider();
+            DataKeyOptions dataKeyOptions = new DataKeyOptions();
+            BsonDocument masterKey = createEncryptedCollectionParams.getMasterKey();
+            if (masterKey != null) {
+                dataKeyOptions.masterKey(masterKey);
+            }
+            String keyIdBsonKey = "keyId";
+            // any mutable non-thread-safe Boolean should do
+            AtomicBoolean dataKeyMayBeCreated = new AtomicBoolean();
+            try {
+                fields.asArray()
+                        .stream()
+                        .filter(BsonValue::isDocument)
+                        .map(BsonValue::asDocument)
+                        .filter(field -> field.containsKey(keyIdBsonKey))
+                        .filter(field -> Objects.equals(field.get(keyIdBsonKey), BsonNull.VALUE))
+                        .forEachOrdered(field -> {
+                            // It is crucial to set the `dataKeyMayBeCreated` flag either immediately before calling `createDataKey`,
+                            // or after that in a `finally` block.
+                            dataKeyMayBeCreated.set(true);
+                            BsonBinary dataKeyId = createDataKey(kmsProvider, dataKeyOptions);
+                            field.put(keyIdBsonKey, dataKeyId);
+                        });
+                database.createCollection(collectionName, createCollectionOptions.clone().encryptedFields(actualEncryptedFields));
+            } catch (RuntimeException e) {
+                if (dataKeyMayBeCreated.get()) {
+                    throw new MongoUpdatedEncryptedFieldsException(actualEncryptedFields, format("Failed to create %s.", namespace), e);
+                } else {
+                    throw e;
```

Comment:
If it happens to be a checked exception, then it must have been thrown sneakily, and I am not sure we should wrap it in a `MongoException`. `RetryingSyncSupplier.get`, updated in [Replace RuntimeException in catch with Exception to support Kotlin #971](https://github.com/mongodb/mongo-java-driver/pull/971) (Maxim linked in a different comment) wraps checked (sneakily thrown) exceptions in `RuntimeException` instead of wrapping them in `MongoException`. I did the same here.